devito/arguments.py

```diff
@@ -1,152 +1,139 @@
-      code generation. Each ArgumentProvider provides one (or more) Argument
-      object(s).
-    - Argument: Classes inheriting from this are for objects that represent the
-      argument itself. Each ArgumentProvider might provide one or more such objects
-      which are used as placeholders for the argument as well as for verification and
-      derivation of default values.
+from devito.tools import filter_ordered, flatten, GenericVisitor
+from devito.function import CompositeFunction, SymbolicFunction
+from devito.dimension import Dimension
+from devito.ir.support.stencil import Stencil
+
+
+""" This module contains a set of classes and functions to deal with runtime arguments
+to Operators. It represents the arguments and their relationships as a DAG (N, E) where
+every node (N) is represented by an object of :class:`Parameter` and every edge is an
+object of class :class:`Dependency`.
+
+The various class hierarchies are explained here:
+Parameter: Any node of the dependency graph has to necessarily be of this type. The node
+           may or may not represent an actual runtime argument passed to the kernel.
+Argument: Subclass of Parameter. This represents a node in the dependency graph directly
+          corresponding to a runtime argument passed to the kernel.
+
+             Parameter
+                 |
+        -------------------
+DimensionParameter        |
+                          |
+                       Argument
+                          |
+                          |
+                --------------------
+                |         |        |
+        ScalarArgument    |        |
+                    TensorArgument |
+                                PtrArgument
+
+ArgumentEngine: The main external API for this module. It encapsulates all the argument
+                derivation and verification logic.
+
+ArgumentVisitor: Visits objects of :class:`function.Basic` types to return appropriate
+                 objects from the above parameter-argument hierarchy.
+
+ValueVisitor: Used by the argument derivation method to derive the value of each parameter
+              based on the dependency tree.
+
+Dependency: Edges of the dependency graph are represented by objects of this class.
+
+UnevaluatedDependency: A "future" object representing an argument derivation that is yet
+                       to happen.
 """
 
 
-class Argument(object):
+class Parameter(object):
+    """ Abstract base class for any object that represents a node in the dependency
+        graph. It may or may not represent a runtime argument.
 
-    """ Abstract base class for any object that represents a run time argument for
-        generated kernels.
+    :param name: Name of the parameter
+    :param dependencies: A list of :class:`Dependency` objects that represent all the
+                         incoming edges into this node of the graph.
     """
-
-    __metaclass__ = abc.ABCMeta
-
+    is_Dimension_Parameter = False
```

Comment:
As a convention, the '_' only lives between "is" and the rest

Reply:
Fixed